### ISSUE TITLE
Deep recursion `PyObject_IsInstance` raises `SystemError`

### ISSUE THREAD

**Issue by nineteendo** [python/cpython#123593]
# Bug report

### Bug description:

I’m recursively calling a function, which in this case checks `PyObject_IsInstance(obj, (PyTypeObject *)s->Mapping)` before repeating. The problem is that when I hit the recursion limit this check raises a `SystemError` instead of a `RecursionError`:

```python
>>> import jsonyx as json # using https://github.com/nineteendo/jsonyx
>>> obj = {}
>>> for _ in range(100_000):
...     obj = {"": obj}
...
>>> json.dump(obj)
Traceback (most recent call last):
  File "<frozen abc>", line 119, in __instancecheck__
RecursionError: maximum recursion depth exceeded in comparison

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Users\wanne\AppData\Local\Programs\Python\Python312\Lib\site-packages\jsonyx\__init__.py", line 614, in dump
    ).dump(obj, fp)
      ^^^^^^^^^^^^^
  File "C:\Users\wanne\AppData\Local\Programs\Python\Python312\Lib\site-packages\jsonyx\__init__.py", line 274, in dump
    fp.write(self._encoder(obj))
             ^^^^^^^^^^^^^^^^^^
SystemError: <built-in method __length_hint__ of dict_keyiterator object at 0x0000023CAC1EF8D0> returned a result with an exception set
```

It can be tested relatively easily by replacing `PyDict_Check()` here:

https://github.com/python/cpython/blob/88210c295d51364c9f779989bc528084b8fe8765/Modules/_json.c#L1423-L1429

### CPython versions tested on:

3.11, 3.12

### Operating systems tested on:

Linux, Windows

**Comment by ZeroIntensity** [comment]
I don't think this is a CPython issue :(

The `SystemError` only occurs if the accelerator module (`_jsonyx`) has been built, which is a very good indicator that this is user-error. In my experience, 99% of the time a `SystemError` is raised, it's the fault of an extension module, and only very rarely CPython. If this were to occur without the extension, then this would be a problem!

It's also worth noting that the `__length_hint__` of `dict_keyiterator` doesn't call any API functions! Therefore, it cannot be the cause of an exception; the problem is the caller. Likewise, the cause seems to be that `jsonyx` lacks *a lot* of error checks. For example, most uses of `_PyUnicodeWriter_WriteChar` are incorrect -- they lack the trailing `< 0` check that would indicate it's an error.

In this case, you forget to take into account that `PyObject_IsInstance` itself can return an exception. The fix is to add a check for that in your condition:

```c
    else if (PyObject_IsInstance(obj, (PyObject *)s->Mapping)) {
        if (PyErr_Occurred()) {
            return -1;
        }
        // ...
    }
```

Adding the above fixes the `SystemError` on my side. (FWIW, I had to change this cast to get this to compile before I did any debugging. Objects passed to `PyObject_IsInstance` should be `PyObject *`, not `PyTypeObject *`.)

I'll submit a PR to your library to fix the missing error checks, and I suggest closing this in the meantime.
